lib/epochs.js

```diff
@@ -267,119 +278,102 @@ function reduceEpochs(ssb, groupId, getters = {}, cb) {
-                  cb(null)
-                })
-              }),
-              pull.collect((err) => {
+      // Take each root/update and build an epoch "node" for our tangle
+      const source = pull(
+        pull.values([
+          pull.values([root]), // TODO pull.once
+          pull(
+            getTangleUpdates.stream(ssb, 'epoch', info.root, { live }),
+            pull.filter(isInitEpoch)
+          ),
+        ]),
+        pull.flatten(),
+        pull.asyncMap((msg, cb) => {
+          const epochRootId = toMsgURI(msg.key)
+          const epochData = {}
+          const node = {
+            key: epochRootId,
+            previous: msg.value.content.tangles.epoch.previous,
+            data: {
+              [epochRootId]: epochData,
+            },
+          }
+          if (!getters) return cb(null, node)
+
+          // Use getters to attach desired data to our epoch "node"
+          pull(
+            pull.values(Object.entries(getters)),
+            pull.asyncMap(([fieldName, getter], cb) => {
+              getter(msg, (err, fieldData) => {
                 // prettier-ignore
-                if (err) return cb(clarify(err, 'Failed to collect epoch data'))
+                if (err) return deferedSource.abort(clarify(err, 'Failed to get epoch ' + fieldName))
 
-                cb(null, node)
+                epochData[fieldName] = fieldData
+                cb(null)
               })
-            )
-          }),
-          pull.collect((err, nodes) => {
-            // prettier-ignore
-            if (err) return cb(clarify(err, 'Failure collecting epoch messages'))
-
-            // Finally reduce all of these epoch nodes with a tangle
-            const reduce = new Reduce(strategy, { nodes })
-            reduce.resolve()
-            // INFO this walks the graph and prunes disconnected nodes
-
-            cb(null, reduce)
-          })
-        )
-      })
+            }),
+            pull.collect((err) => {
+              // prettier-ignore
+              if (err) return deferedSource.abort(clarify(err, 'Failed to collect epoch data'))
+
+              cb(null, node)
+            })
+          )
+        })
+      )
+      deferedSource.resolve(source)
     })
   })
-}
 
-function getMemberUpdates(ssb, epochRootId, opts) {
-  return pull(
-    getTangleUpdates.stream(ssb, 'members', epochRootId, {
-      live: !!opts?.live,
-    }),
-    pull.filter((msg) => isAddMember(msg) || isExclude(msg)),
-    pull.map((msg) => {
-      const { type, recps, excludes } = msg.value.content
-      if (type === 'group/add-member')
-        return recps.slice(1).map((feedId) => ({ added: feedId }))
-      else return excludes.map((feedId) => ({ excluded: feedId }))
-    }),
-    pull.flatten(),
-    // grab updates until we get an exclusion, then close the stream (even if live)
-    // last: true includes the exclusion as well right before closing
-    pull.take((update) => update.added, { last: true })
```

Comment:
i see you moved this function to `getMembers.stream`, that's fine

but i think this line was important, it was part of a bugfix. i suspect the last listMembers test will fail without it

Reply:
true enough! I found another solution but I think this is probably more elegant ... lemme look

Reply:
https://github.com/ssbc/ssb-tribes2/pull/104#discussion_r1183642865

p.s. I didn't know you could use pull.take like this. Awesome